File: server/src/shared/helpers/result.ts
export class Result<T> {
  success: boolean;
  data?: T;
  error?: string;

  private constructor(success: boolean, data?: T, error?: string) {
    this.success = success;
    this.data = data;
    this.error = error;
  }

  static success<T>(data?: T) {
    return new Result(true, data);
  }

  static failure<T>(error: string, data?: T) {
    return new Result(false, data, error);
  }
}

File: client/lib/shared/helpers/result.dart
class Result<T> {
  final bool success;
  final T? data;
  final String? error;

  Result._({required this.success, this.data, this.error});

  factory Result.success([T? data]) {
    return Result._(success: true, data: data);
  }

  factory Result.failure(String error, [T? data]) {
    return Result._(success: false, data: data, error: error);
  }
}
